### PR TITLE
fix(lwm2m): erase the trans when timeout

### PR DESCRIPTION
**Pull request by terry-xiaoyu**
https://github.com/emqx/emqx/issues/4198

### CODE REVIEW

File: src/lwm2m_coap_channel.erl [hunk 1]
```diff
@@ -301,8 +301,8 @@ next_mid(MsgId) ->
         true -> 1 % or 0?
     end.
 
-update_state(State=#state{trans=Trans}, _TrId, undefined) ->
-    {noreply, State#state{trans=Trans}, hibernate};
+update_state(State=#state{trans = Trans0}, TrId, undefined) ->
```
Comment (zmstone) [inline src/lwm2m_coap_channel.erl:304]: sorry, I'm not quite following.
if this is intended for the case of this caller:
```
handle_info({timeout, TrId, Event}, State=#state{trans=Trans}) ->
    update_state(State, TrId,
        case dict:find(TrId, Trans) of
            error -> undefined; % ignore unexpected responses
            {ok, TrState} -> lwm2m_coap_transport:timeout(Event, TrState)
        end);
```
then it's nothing to erase isn't it ?

Comment (terry-xiaoyu) [reply src/lwm2m_coap_channel.erl:304]: The `lwm2m_coap_transport:timeout/2` always returns `undefined`. I don't know why the original writer do this, but it should be erased when the transport is timeout.

I've refactored a lot in the lwm2m/coap at China Telecom when I was doing customization work, but in their private repo. We may need to merge some of the code to open-source lwm2m here.

Comment (zmstone) [reply src/lwm2m_coap_channel.erl:304]: if we do not have any other place which calls update_state with an `undefined` atom as the third arg,
I'd suggest that we do dict:erase here directly, it should be much easier to reason about.

Comment (terry-xiaoyu) [reply src/lwm2m_coap_channel.erl:304]: Then we have to match the {timeout, transport} in the lwm2m_coap_channel module, delete the trans state in this case. Or we need to get the return value of `lwm2m_coap_transport:timeout/2` and delete the state only when it returns undefined.

The first way we would expose details of transport in to the outer module, and the second needs a little more change to the original code than mine PR suggested.

**Pull request merged**